File: benchmarks/lang/lua/fannkuch_redux.lua
-- Fannkuch-redux. 0-indexed port of Mike Pall's algorithm, matching the
-- Tzopilotl version line-for-line so the comparison is apples-to-apples.
local N = 10

local function fannkuch(n)
    local p, q, s = {}, {}, {}
    for i = 0, n - 1 do p[i] = i; q[i] = i; s[i] = i end

    local sign = 1
    local maxflips = 0
    local sum = 0

    while true do
        for j = 0, n - 1 do q[j] = p[j] end
        local flips = 0
        local q0 = q[0]
        while q0 ~= 0 do
            local lo, hi = 0, q0
            while lo < hi do
                q[lo], q[hi] = q[hi], q[lo]
                lo = lo + 1
                hi = hi - 1
            end
            flips = flips + 1
            q0 = q[0]
        end
        if flips > maxflips then maxflips = flips end
        sum = sum + sign * flips

        if sign > 0 then
            p[0], p[1] = p[1], p[0]
            sign = -1
        else
            p[1], p[2] = p[2], p[1]
            sign = 1
            local i = 2
            local stop = false
            while i < n do
                local si = s[i]
                if si ~= 0 then
                    s[i] = si - 1
                    break
                end
                if i == n - 1 then stop = true; break end
                s[i] = i
                local t2 = p[0]
                for k = 0, i do p[k] = p[k + 1] end
                p[i + 1] = t2
                i = i + 1
            end
            if stop then break end
        end
    end
    return sum, maxflips
end

local sum, maxflips = fannkuch(N)
print(sum)
print("Pfannkuchen(" .. N .. ") = " .. maxflips)

File: benchmarks/lang/cpp/fannkuch_redux.cpp
// Fannkuch-redux. 0-indexed port of Mike Pall's algorithm, matching the
// Lua/Tzopilotl versions line-for-line so the comparison is apples-to-apples.
#include <cstdio>

static const int N = 10;

static void fannkuch(int n, long long& out_sum, long long& out_maxflips) {
    int p[64], q[64], s[64];
    for (int i = 0; i < n; ++i) { p[i] = i; q[i] = i; s[i] = i; }

    int sign = 1;
    long long maxflips = 0;
    long long sum = 0;

    while (true) {
        for (int j = 0; j < n; ++j) q[j] = p[j];
        long long flips = 0;
        int q0 = q[0];
        while (q0 != 0) {
            int lo = 0, hi = q0;
            while (lo < hi) {
                int tmp = q[lo]; q[lo] = q[hi]; q[hi] = tmp;
                ++lo; --hi;
            }
            ++flips;
            q0 = q[0];
        }
        if (flips > maxflips) maxflips = flips;
        sum += sign * flips;

        if (sign > 0) {
            int t = p[0]; p[0] = p[1]; p[1] = t;
            sign = -1;
        } else {
            int t = p[1]; p[1] = p[2]; p[2] = t;
            sign = 1;
            int i = 2;
            bool stop = false;
            while (i < n) {
                int si = s[i];
                if (si != 0) { s[i] = si - 1; break; }
                if (i == n - 1) { stop = true; break; }
                s[i] = i;
                int t2 = p[0];
                for (int k = 0; k <= i; ++k) p[k] = p[k + 1];
                p[i + 1] = t2;
                ++i;
            }
            if (stop) break;
        }
    }
    out_sum = sum;
    out_maxflips = maxflips;
}

int main() {
    long long sum, maxflips;
    fannkuch(N, sum, maxflips);
    std::printf("%lld\n", sum);
    std::printf("Pfannkuchen(%d) = %lld\n", N, maxflips);
}
